File: app/src/main/java/com/github/fo2rist/mclaren/repository/converters/StoryStreamConverter.java
package com.github.fo2rist.mclaren.repository.converters;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Html;
import android.text.TextUtils;

import com.github.fo2rist.mclaren.models.FeedItem;
import com.github.fo2rist.mclaren.models.FeedItem.SourceType;
import com.github.fo2rist.mclaren.models.FeedItem.Type;
import com.github.fo2rist.mclaren.models.ImageUrl;
import com.github.fo2rist.mclaren.models.Size;
import com.github.fo2rist.mclaren.web.models.StoryStream;
import com.github.fo2rist.mclaren.web.models.StoryStreamItem;
import com.github.fo2rist.mclaren.web.models.StoryStreamItemWrapper;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import static com.github.fo2rist.mclaren.models.FeedItem.TEXT_LENGTH_LIMIT;
import static com.github.fo2rist.mclaren.web.models.StoryStreamItem.FeedType;
import static com.github.fo2rist.mclaren.web.models.StoryStreamItem.ImageData;
import static com.github.fo2rist.mclaren.web.models.StoryStreamItem.VideoData;

/**
 * Converts StoryStream API feed web-model to app models.
 */
public final class StoryStreamConverter implements FeedConverter<StoryStream> {

    public static final StoryStreamConverter INSTANCE = new StoryStreamConverter();

    private static final String HTTP = "http";

    private StoryStreamConverter() {
    }

    @NonNull
    @Override
    public List<FeedItem> convertFeed(@NonNull StoryStream storyStreamFeed) {
        ArrayList<FeedItem> result = new ArrayList<>(storyStreamFeed.items.size());
        for (StoryStreamItemWrapper storyStreamItem: storyStreamFeed.items) {
            // TODO: Parser may fail and produce useless feed item -> Filter out bad items. 2019-07-31
            result.add(convertFeedItem(storyStreamItem));
        }
        return result;
    }
    
    private static FeedItem convertFeedItem(StoryStreamItemWrapper storyStreamItem) {
        return new FeedItem(
            fetchId(storyStreamItem),
            fetchType(storyStreamItem),
            fetchText(storyStreamItem),
            fetchContent(storyStreamItem),
            fetchDate(storyStreamItem),
            fetchSourceType(storyStreamItem),
            fetchSourceName(storyStreamItem),
            fetchMediaLink(storyStreamItem),
            fetchImageUrls(storyStreamItem));
    }

    private static long fetchId(@NonNull StoryStreamItemWrapper storyStreamItem) {
        //StoryStream don't offer sequential IDs so use timestamp instead as a hack
        Date publishDate = storyStreamItem.publishDate;
        return publishDate != null
                ? publishDate.getTime()
                : 0; // TODO default values makes no sense need. Check if this kind of item can be displayed. 2019-07-31
    }

    @NonNull
    private static Type fetchType(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);

        if (storyStreamContentItem == null) {
            //TODO Default value doesn't reflect the actual intent in case the content item is null. 2019-07-31
            return Type.Message;
        }

        if (storyStreamContentItem.feedType == FeedType.Custom) {
            return Type.Article;
        } else {
            if (storyStreamContentItem.contentType == null) {
                return Type.Message;
            }
            switch (storyStreamContentItem.contentType) {
                case Text:
                    return Type.Message;
                case Image:
                    if (storyStreamContentItem.images.size() > 1) {
                        return Type.Gallery;
                    } else {
                        return Type.Image;
                    }
                case Video:
                    return Type.Video;
                default:
                    return Type.Message;
            }
        }
    }

    @NonNull
    private static String fetchText(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);
        if (storyStreamContentItem == null) {
            return "";
        }
        if (storyStreamContentItem.body == null) {
            return "";
        }

        if (storyStreamContentItem.feedType == FeedType.Custom) {
            String text = storyStreamContentItem.body;
            text = Html.fromHtml(text).toString();

            text = text.replaceAll("\\n+", " ");

            if (!TextUtils.isEmpty(storyStreamContentItem.title)) {
                text = storyStreamContentItem.title.toUpperCase(Locale.US) + "\n" + text;
            }

            if (text.length() > TEXT_LENGTH_LIMIT) {
                text = text.substring(0, TEXT_LENGTH_LIMIT) + "...";
            }

            return text;
        } else {
            //TODO We are loosing links that way. HTML should be preserved and then handled on the UI side. 2018.03.09
            return Html.fromHtml(storyStreamContentItem.body).toString();
        }
    }

    @Nullable
    private static StoryStreamItem fetchContentItem(@NonNull StoryStreamItemWrapper storyStreamItem) {
        ArrayList<StoryStreamItem> contentItems = storyStreamItem.contentItems;
        return contentItems != null
                ? contentItems.get(0)
                : null;
    }

    @Nullable
    private static String fetchContent(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);
        return storyStreamContentItem != null
                ? storyStreamContentItem.body
                : null;
    }

    @NonNull
    private static Date fetchDate(@NonNull StoryStreamItemWrapper storyStreamItem) {
        return storyStreamItem.publishDate != null
                ? storyStreamItem.publishDate
                : new Date(); //TODO default value makes no sense. Check if it can be used anyhow. 2019-07-31
    }

    @NonNull
    private static SourceType fetchSourceType(@NonNull StoryStreamItemWrapper storyStreamItem) {
        FeedType feedType = fetchContentItem(storyStreamItem).feedType;
        if (feedType == null) {
            return SourceType.Unknown;
        }
        switch(feedType) {
            case Twitter:
                return SourceType.Twitter;
            case Instagram:
                return SourceType.Instagram;
            default:
                return SourceType.Unknown;
        }
    }

    @NonNull
    private static String fetchSourceName(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);
        //for pretty name use .author not .source
        return (storyStreamContentItem != null && storyStreamContentItem.source != null)
                ? storyStreamContentItem.source
                : "" ;
    }

    @NonNull
    private static String fetchMediaLink(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);
        if (storyStreamContentItem == null) {
            return "";
        }
        List<VideoData> videos = storyStreamContentItem.videos;
        return (videos != null && !videos.isEmpty() && videos.get(0) != null)
                ? videos.get(0).url
                : "";
    }

    @NonNull
    private static List<ImageUrl> fetchImageUrls(@NonNull StoryStreamItemWrapper storyStreamItem) {
        StoryStreamItem storyStreamContentItem = fetchContentItem(storyStreamItem);
        if (storyStreamContentItem == null) {
            return new ArrayList<>();
        }

        List<ImageData> images = storyStreamContentItem.images;
        if (images == null) {
            return new ArrayList<>();
        }

        List<ImageUrl> result = new ArrayList<>(images.size());
        for (int i = 0; i < images.size(); i++) {
            result.add(fetchUrlFromImageData(images.get(i)));
        }
        return result;
    }

    private static ImageUrl fetchUrlFromImageData(@NonNull ImageData imageData) {
        String originalSizeUrl = fixUrl(imageData.originalSizeUrl);
        Size originalSize = toImageSize(imageData.sizes.originalSize);

        if (!originalSizeUrl.isEmpty()) {
            //Here the original link is present additional links are optional

            String twoUpSizeUrl = fixUrl(imageData.twoUpSizeUrl);
            String threeUpSizeUrl = fixUrl(imageData.threeUpSizeUrl);

            if (originalSizeUrl.equals(twoUpSizeUrl) && originalSizeUrl.equals(threeUpSizeUrl)) {
                //Broken links are usually equal, so we should ignore small size links as incorrect
                return ImageUrl.create(originalSizeUrl, originalSize);
            } else if (twoUpSizeUrl.isEmpty() && threeUpSizeUrl.isEmpty()){
                //We only have one link of three, create a single URL
                return ImageUrl.create(originalSizeUrl, originalSize);
            } else {
                //Normal case - all three links are different and not empty
                Size twoUpSize = toImageSize(imageData.sizes.twoUpSize);
                Size threeUpSize = toImageSize(imageData.sizes.threeUpSize);
                return ImageUrl.create(originalSizeUrl, originalSize,
                        twoUpSizeUrl, twoUpSize,
                        threeUpSizeUrl, threeUpSize);
            }
        } else if (imageData.name.startsWith(HTTP)) {
            //Here original link is absent

            //when links are broken the name usually contains the link
            return ImageUrl.create(imageData.name, originalSize);
        } else {
            //nothing we can do
            return ImageUrl.empty();
        }
    }

    /**
     * Strip URL from non-http prefix.
     * Sometimes API return weird prefix before the actual URL, which should be stripped.
     * @return empty string if input is null or empty, fixed URL otherwise.
     */
    @NonNull
    private static String fixUrl(@Nullable String url) {
        if (url == null) {
            return "";
        }

        if (!url.startsWith(HTTP) && url.contains(HTTP) ) {
            return url.replaceFirst("^.*\\/http?", "http");
        } else {
            return url;
        }
    }

    @NonNull
    private static Size toImageSize(@Nullable StoryStreamItem.ImageSize size) {
        if (size == null) {
            return Size.UNKNOWN;
        }
        return Size.valueOf(size.width, size.height);
    }
}

File: app/src/main/java/com/github/fo2rist/mclaren/repository/converters/StoryStreamConverter.kt
package com.github.fo2rist.mclaren.repository.converters

import android.text.Html
import com.github.fo2rist.mclaren.models.FeedItem
import com.github.fo2rist.mclaren.models.FeedItem.Companion.TEXT_LENGTH_LIMIT
import com.github.fo2rist.mclaren.models.FeedItem.SourceType
import com.github.fo2rist.mclaren.models.FeedItem.Type
import com.github.fo2rist.mclaren.models.ImageUrl
import com.github.fo2rist.mclaren.models.Size
import com.github.fo2rist.mclaren.web.models.StoryStream
import com.github.fo2rist.mclaren.web.models.StoryStreamItem
import com.github.fo2rist.mclaren.web.models.StoryStreamItem.ContentType
import com.github.fo2rist.mclaren.web.models.StoryStreamItem.FeedType
import com.github.fo2rist.mclaren.web.models.StoryStreamItem.ImageData
import com.github.fo2rist.mclaren.web.models.StoryStreamItem.ImageSize
import com.github.fo2rist.mclaren.web.models.StoryStreamWrappingItem
import java.util.Date
import java.util.Locale

private const val HTTP = "http"

/**
 * Converts StoryStream API feed web-model to app models.
 */
object StoryStreamConverter : FeedConverter<StoryStream> {

    override fun convertFeed(feedWebModel: StoryStream): List<FeedItem> {
        return feedWebModel.items?.mapNotNull {
            convertFeedItem(it)
        } ?: emptyList()
    }

    private fun convertFeedItem(storyStreamWrappingItem: StoryStreamWrappingItem): FeedItem? {
        val contentItem = fetchContentItem(storyStreamWrappingItem)
                ?: return null

        return FeedItem(
                fetchId(contentItem),
                fetchType(contentItem),
                fetchText(contentItem),
                fetchContent(contentItem),
                fetchDate(contentItem),
                fetchSourceType(contentItem),
                fetchSourceName(contentItem),
                fetchMediaLink(contentItem),
                fetchImageUrls(contentItem))
    }

    private fun fetchContentItem(storyStreamItem: StoryStreamWrappingItem): StoryStreamItem? {
        return storyStreamItem.contentItems?.firstOrNull()
    }

    private fun fetchId(storyStreamItem: StoryStreamItem): Long {
        //StoryStream don't offer sequential IDs so use timestamp instead as a hack
        val publishDate = storyStreamItem.publishDate
        return publishDate?.time
                ?: 0 // TODO default values makes no sense need. Check if this kind of item can be displayed. 2019-07-31
    }

    private fun fetchType(storyStreamItem: StoryStreamItem): Type {
        if (storyStreamItem.feedType == FeedType.Custom) {
            return Type.Article
        }

        return when {
            storyStreamItem.contentType == ContentType.Text ->
                Type.Message
            storyStreamItem.contentType == ContentType.Video ->
                Type.Video
            storyStreamItem.contentType == ContentType.Image && storyStreamItem.images.len == 1 ->
                Type.Image
            storyStreamItem.contentType == ContentType.Image && storyStreamItem.images.len > 1 ->
                Type.Gallery
            else ->
                Type.Message
        }
    }

    private val List<*>?.len: Int
        get() = this?.size ?: 0

    @Suppress("DEPRECATION") //Html.fromHtml is deprecated but new version is only available on SDK 24
    private fun fetchText(storyStreamItem: StoryStreamItem): String {
        val rawBody = storyStreamItem.body
                ?: return ""

        if (storyStreamItem.feedType == FeedType.Custom) {
            val textBody = Html.fromHtml(rawBody)
                    .toString()
                    .replace("\\n+".toRegex(), " ")

            val textTitle = storyStreamItem.title
            val text = if (textTitle.isNullOrEmpty()) {
                textBody
            } else {
                textTitle.toUpperCase(Locale.US) + "\n" + textBody
            }

            return text.trimmed()
        } else {
            //TODO We are loosing links that way. HTML should be preserved and then handled on the UI side. 2018.03.09
            return Html.fromHtml(rawBody).toString()
        }
    }

    private fun String.trimmed(): String {
        return if (this.length > TEXT_LENGTH_LIMIT) {
            this.substring(0, TEXT_LENGTH_LIMIT) + "..."
        } else {
            this
        }
    }

    private fun fetchContent(storyStreamItem: StoryStreamItem): String? {
        return storyStreamItem.body
    }

    private fun fetchDate(storyStreamItem: StoryStreamItem): Date {
        return storyStreamItem.publishDate
                ?: Date() //TODO default value makes no sense. Check if it can be used anyhow. 2019-07-31
    }

    private fun fetchSourceType(storyStreamItem: StoryStreamItem): SourceType {
        return when (storyStreamItem.feedType) {
            FeedType.Twitter -> SourceType.Twitter
            FeedType.Instagram -> SourceType.Instagram
            else -> SourceType.Unknown
        }
    }

    private fun fetchSourceName(storyStreamItem: StoryStreamItem): String {
        //for pretty name use .author not .source
        return storyStreamItem.source
                ?: ""
    }

    private fun fetchMediaLink(storyStreamItem: StoryStreamItem): String {
        return storyStreamItem.videos?.firstOrNull()?.url
                ?: ""
    }

    private fun fetchImageUrls(storyStreamItem: StoryStreamItem): List<ImageUrl> {
        val images = storyStreamItem.images
                ?: return emptyList()

        return images.map {
            fetchUrlFromImageData(it)
        }
    }

    private fun fetchUrlFromImageData(imageData: ImageData): ImageUrl {
        val originalSizeUrl = fixUrl(imageData.originalSizeUrl)
        val originalSize = imageData.sizes?.originalSize.toImageSize()

        if (originalSizeUrl.isNotEmpty()) {
            //Here the original link is present additional links are optional

            val twoUpSizeUrl = fixUrl(imageData.twoUpSizeUrl)
            val threeUpSizeUrl = fixUrl(imageData.threeUpSizeUrl)

            val twoUpSize = imageData.sizes?.twoUpSize.toImageSize()
            val threeUpSize = imageData.sizes?.threeUpSize.toImageSize()

            return buildImageUrlFromMultipleUrls(
                    originalSizeUrl, twoUpSizeUrl, threeUpSizeUrl,
                    originalSize, twoUpSize, threeUpSize)
        } else {
            //Here original link is absent
            //when links are broken the name usually contains the link
            return buildImageUrlFromName(imageData.name, originalSize)
        }
    }

    @Suppress("Detekt.LongParameterList")
    private fun buildImageUrlFromMultipleUrls(
        originalSizeUrl: String, twoUpSizeUrl: String, threeUpSizeUrl: String,
        originalSize: Size, twoUpSize: Size, threeUpSize: Size
    ): ImageUrl {
        return if (originalSizeUrl == twoUpSizeUrl && originalSizeUrl == threeUpSizeUrl) {
            //Broken links are usually equal, so we should ignore small size links as incorrect
            ImageUrl.create(originalSizeUrl, originalSize)
        } else if (twoUpSizeUrl.isEmpty() && threeUpSizeUrl.isEmpty()) {
            //We only have one link of three, create a single URL
            ImageUrl.create(originalSizeUrl, originalSize)
        } else {
            //Normal case - all three links are different and not empty
            ImageUrl.create(originalSizeUrl, originalSize,
                    twoUpSizeUrl, twoUpSize,
                    threeUpSizeUrl, threeUpSize)
        }
    }

    private fun buildImageUrlFromName(imageName: String?, originalSize: Size): ImageUrl {
        return if (imageName != null && imageName.startsWith(HTTP)) {
            ImageUrl.create(imageName, originalSize)
        } else {
            //if it doesn't there is nothing we can do
            ImageUrl.empty()
        }
    }

    /**
     * Strip URL from non-http prefix.
     * Sometimes API return weird prefix before the actual URL, which should be stripped.
     * @return empty string if input is null, empty, or doesn't contain URL
     */
    private fun fixUrl(url: String?): String {
        return when {
            url == null -> ""
            !url.startsWith(HTTP) && url.contains(HTTP) -> url.replaceFirst("^.*\\/http?".toRegex(), HTTP)
            else -> url
        }
    }

    private fun ImageSize?.toImageSize(): Size {
        return if (this == null) {
            Size.UNKNOWN
        } else {
            Size.valueOf(width, height)
        }
    }
}
